include/RAJA/policy/hip/policy.hpp

```diff
@@ -396,21 +333,190 @@ struct HipDimHelper<2>{
 template<int dim, typename dim_t>
 RAJA_HOST_DEVICE
 constexpr
-auto get_hip_dim(dim_t const &d) ->
-  decltype(HipDimHelper<dim>::get(d))
+auto get_hip_dim(dim_t const &d)
 {
   return HipDimHelper<dim>::get(d);
 }
 
 template<int dim, typename dim_t>
 RAJA_HOST_DEVICE
-void set_hip_dim(dim_t &d, int value)
+void set_hip_dim(dim_t &d, HipDimIdxT value)
 {
   return HipDimHelper<dim>::set(d, value);
 }
+
+
+/// Type representing thread indexing within a block
+/// block_size is fixed
+template<int dim, HipDimIdxT t_block_size>
+struct HipIndexThread {
+
+  template < typename IdxT = HipDimIdxT >
+  RAJA_DEVICE
+  static inline constexpr
+  auto index() { return static_cast<IdxT>(HipDimHelper<dim>::get(threadIdx)); }
+
+  template < typename IdxT = HipDimIdxT >
+  RAJA_DEVICE
+  static inline constexpr
+  auto size() { return static_cast<IdxT>(t_block_size); }
+};
+/// unless t_block_size is 0 then block_size is dynamic
+template<int dim>
+struct HipIndexThread<dim, 0> {
```

Comment:
All of the Hip Indexing allows you to put a template param number which is used as the size. If that template param is 0 then a runtime size is used instead (like blockDim.x). It is an error if the template param specifies a non-zero size but the block/grid does not have that size.